### PR TITLE
ByteString serialization

### PR DESCRIPTION
**Pull request by colinrgodsey**
Addresses ByteString serialization (#15170) while keeping binary
compatibility with the previously serializable ByteString1C

Squashed from previous PR ~~#17070~~

Referenced in previous ByteString comments in #17028

### CODE REVIEW

**Comment by drewhk** [comment]
OK TO TEST

**Comment by ktoso** [comment]
LGTM, was reviewed in the previous PR already.

@colinrgodsey please adjust the commit message according to our guidelines [CONTRIBUTING.md](https://github.com/akka/akka/blob/master/CONTRIBUTING.md), in this case it would be: `=act #15170 bytestring serialization` Thanks! (You can simply `git commit --amen -m '=act #15170 bytestring serialization' && git push -f` to update this PR)

**Comment by colinrgodsey** [comment]
@ktoso wow, didnt even know about force....

fixed.

**Comment by patriknw** [comment]
LGTM

**Comment by ktoso** [comment]
@colinrgodsey, yeah you can also `push -f` after squashing commits etc, so your work could have remained in the initial PR :-) Just a hint for the future, not a problem at all.

Thanks for the contrib again!

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 1]
```diff
@@ -274,8 +322,43 @@ object ByteString {
...
+      val serializationId = in.readByte()
+
+      orig = Companion(from = serializationId).readFromInputStream(in)
+    }
+
+    private def readResolve(): AnyRef = orig
+  }
+
+  private[akka] object Companion {
+    private val companionMap = Seq(ByteString1, ByteString1C, ByteStrings).
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: Wouldn't this be much more performant with a `match` + `@tableswitch`?

Comment (colinrgodsey) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: hmm, probably shouldve =\

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: How about submitting a PR with that improvement @colinrgodsey? Would be very welcome :-)

**Pull request merged**